### ISSUE TITLE
Internal error occurred: failed calling webhook "mtargetgroupbinding.elbv2.k8s.aws" (calico not involved!)

### ISSUE THREAD

**Issue by crielly** [kubernetes-sigs/aws-load-balancer-controller#1634]
Some basic info:
EKS 1.18
Using default CNI

I'm trying to register a pod as an ALB target using https://github.com/kubernetes-sigs/aws-load-balancer-controller/releases/tag/v2.0.0

The error I'm seeing is identical to https://github.com/kubernetes-sigs/aws-load-balancer-controller/issues/1591 - when I try to register the TargetGroupBinding:

```apiVersion: elbv2.k8s.aws/v1beta1
kind: TargetGroupBinding
metadata:
  name: 2048-game-tgb
spec:
  serviceRef:
    name: game-2048 # route traffic to the awesome-service
    port: 80
  targetGroupARN: arn:aws:elasticloadbalancing:us-west-2:<someAccountID>:targetgroup/qa-game/8b1e5ee6c2568eb9
```

Using a basic demo app to test the ingress, I get

```service/service-2048 unchanged
Error from server (InternalError): error when creating "/home/crielly/Desktop/aws-lb-controller-yaml/2048.yaml": Internal error occurred: failed calling webhook "mtargetgroupbinding.elbv2.k8s.aws": Post https://aws-load-balancer-webhook-service.kube-system.svc:443/mutate-elbv2-k8s-aws-v1beta1-targetgroupbinding?timeout=30s: context deadline exceeded
```

If I understand correctly, the kube-apiserver is trying to contact my aws-lb-controller service. I've confirmed that 
* DNS seems to be resolving (pod-to-pod at least)
* the SSL handshake SHOULD be functioning (Certs match)
* SGs and NACLs aren't blocking the request: `172.21.16.140 172.21.20.129 443 39332 6 71 42442 1605037839 1605037854 ACCEPT OK` - VPC flow logs show packets from the control plane ENIs being accepted

And just for good measure, I've enabled `hostNetwork: true` as suggested for calico in the above linked ticket. I'm stumped and don't know where to go from here for debugging. Halp please!

**Comment by kishorj** [comment]
I want to rule out whether you had the webhooks from the v2.0.0-rc versions. Also share any errors in the aws-load-balancer-controller logs.

**Comment by crielly** [comment]
@kishorj I've ensured that I'm using the latest releases of cert-manager and the LB controller yaml manifests.

**Comment by crielly** [comment]
@kishorj Logs from the LB controller pod:
```2020-11-11T17:44:36.107931957Z {"level":"info","ts":1605116676.1077864,"msg":"version","GitVersion":"v2.0.0-dirty","GitCommit":"1028fa4f363a9a8e37b07ff6a093b7b422923512","BuildDate":"2020-10-23T01:34:38+0000"}
2020-11-11T17:44:36.207676106Z {"level":"info","ts":1605116676.2075505,"logger":"controller-runtime.metrics","msg":"metrics server is starting to listen","addr":":8080"}
2020-11-11T17:44:36.209028833Z {"level":"info","ts":1605116676.2089515,"logger":"setup","msg":"adding health check for controller"}
2020-11-11T17:44:36.209106033Z {"level":"info","ts":1605116676.2090206,"logger":"controller-runtime.webhook","msg":"registering webhook","path":"/mutate-v1-pod"}
2020-11-11T17:44:36.209113764Z {"level":"info","ts":1605116676.209052,"logger":"controller-runtime.webhook","msg":"registering webhook","path":"/mutate-elbv2-k8s-aws-v1beta1-targetgroupbinding"}
2020-11-11T17:44:36.209118564Z {"level":"info","ts":1605116676.2090776,"logger":"controller-runtime.webhook","msg":"registering webhook","path":"/validate-elbv2-k8s-aws-v1beta1-targetgroupbinding"}
2020-11-11T17:44:36.209312185Z {"level":"info","ts":1605116676.2092564,"logger":"setup","msg":"starting podInfo repo"}
2020-11-11T17:44:38.210197114Z {"level":"info","ts":1605116678.209487,"logger":"controller-runtime.manager","msg":"starting metrics server","path":"/metrics"}
2020-11-11T17:44:38.210224994Z I1111 17:44:38.209562       1 leaderelection.go:242] attempting to acquire leader lease  kube-system/aws-load-balancer-controller-leader...
2020-11-11T17:44:38.309871783Z {"level":"info","ts":1605116678.309679,"logger":"controller-runtime.webhook.webhooks","msg":"starting webhook server"}
2020-11-11T17:44:38.310223993Z {"level":"info","ts":1605116678.310141,"logger":"controller-runtime.certwatcher","msg":"Updated current TLS certificate"}
2020-11-11T17:44:38.310272103Z {"level":"info","ts":1605116678.3102355,"logger":"controller-runtime.webhook","msg":"serving webhook server","host":"","port":9443}
2020-11-11T17:44:38.310398304Z {"level":"info","ts":1605116678.31034,"logger":"controller-runtime.certwatcher","msg":"Starting certificate watcher"}
2020-11-11T17:44:56.750664912Z I1111 17:44:56.750550       1 leaderelection.go:252] successfully acquired lease kube-system/aws-load-balancer-controller-leader
2020-11-11T17:44:56.751044673Z {"level":"info","ts":1605116696.750931,"logger":"controller","msg":"Starting EventSource","reconcilerGroup":"elbv2.k8s.aws","reconcilerKind":"TargetGroupBinding","controller":"targetGroupBinding","source":"kind source: /, Kind="}
2020-11-11T17:44:56.751174284Z {"level":"info","ts":1605116696.7511182,"logger":"controller","msg":"Starting EventSource","reconcilerGroup":"elbv2.k8s.aws","reconcilerKind":"TargetGroupBinding","controller":"targetGroupBinding","source":"kind source: /, Kind="}
2020-11-11T17:44:56.751254135Z {"level":"info","ts":1605116696.751175,"logger":"controller","msg":"Starting EventSource","controller":"ingress","source":"channel source: 0xc0004911d0"}
2020-11-11T17:44:56.751314507Z {"level":"info","ts":1605116696.7512724,"logger":"controller","msg":"Starting EventSource","controller":"ingress","source":"channel source: 0xc000491220"}
2020-11-11T17:44:56.751429901Z {"level":"info","ts":1605116696.7513597,"logger":"controller","msg":"Starting EventSource","controller":"ingress","source":"kind source: /, Kind="}
2020-11-11T17:44:56.751512173Z {"level":"info","ts":1605116696.751454,"logger":"controller","msg":"Starting EventSource","controller":"service","source":"kind source: /, Kind="}
2020-11-11T17:44:56.751663532Z {"level":"info","ts":1605116696.7516131,"logger":"controller","msg":"Starting EventSource","controller":"ingress","source":"kind source: /, Kind="}
2020-11-11T17:44:56.751775991Z {"level":"info","ts":1605116696.7517302,"logger":"controller","msg":"Starting EventSource","controller":"ingress","source":"kind source: /, Kind="}
2020-11-11T17:44:56.751888865Z {"level":"info","ts":1605116696.751848,"logger":"controller","msg":"Starting Controller","controller":"service"}
2020-11-11T17:44:56.851883158Z {"level":"info","ts":1605116696.851744,"logger":"controller","msg":"Starting EventSource","reconcilerGroup":"elbv2.k8s.aws","reconcilerKind":"TargetGroupBinding","controller":"targetGroupBinding","source":"kind source: /, Kind="}
2020-11-11T17:44:56.852245627Z {"level":"info","ts":1605116696.8521762,"logger":"controller","msg":"Starting Controller","controller":"ingress"}
2020-11-11T17:44:56.852564804Z {"level":"info","ts":1605116696.852428,"logger":"controller","msg":"Starting workers","controller":"service","worker count":3}
2020-11-11T17:44:56.952505501Z {"level":"info","ts":1605116696.9522414,"logger":"controller","msg":"Starting Controller","reconcilerGroup":"elbv2.k8s.aws","reconcilerKind":"TargetGroupBinding","controller":"targetGroupBinding"}
2020-11-11T17:44:56.952529689Z {"level":"info","ts":1605116696.9522805,"logger":"controller","msg":"Starting workers","reconcilerGroup":"elbv2.k8s.aws","reconcilerKind":"TargetGroupBinding","controller":"targetGroupBinding","worker count":3}
2020-11-11T17:44:56.952536065Z {"level":"info","ts":1605116696.9523568,"logger":"controller","msg":"Starting workers","controller":"ingress","worker count":3}
2020-11-11T17:51:16.133907317Z 2020/11/11 17:51:16 http: TLS handshake error from 172.21.21.88:48474: local error: tls: bad record MAC```

That last entry, the TLS handshake error - that's me trying to POST the URL above from another unrelated pod, which reports
```curl: (60) SSL certificate problem: unable to get local issuer certificate
More details here: https://curl.haxx.se/docs/sslcerts.html```

Unsure if that is relevant. It seems as though the request from the control plane is not being logged.

**Comment by kishorj** [comment]
could you share your cluster ARN? You can email it to joshikis AT amazon.com, or DM me on kubernetes slack.

**Comment by crielly** [comment]
My apologies - this turned out to not be what I thought it was.

When trying to create my TargetGroupBinding, I was failing to specify spec.targetType. From the docs:

`If TargetType is not explicitly specified, a mutating webhook will automatically call AWS API to find the TargetType for your TargetGroup and set it to correct value.`

This step was failing. I manually specified the targetType matching my LB target group and that appeared to get me past the issue. However, based on that line from the docs above, I didn't really solve the underlying issue which should allow the controller to figure out the target type for me.

@kishorj If you think there's an underlying bug here I'm happy to help with whatever information would be useful - however, I'm no longer blocked here.

**Comment by crielly** [comment]
Even once I manually specified the targetType, I ran into another issue which ultimately helped me figure out the root cause here. When the AWS controller pod tried to bind my endpoints to the specified target group, it was failing to make the API call to AWS due to  `"error":"WebIdentityErr: failed to retrieve credentials\ncaused by: InvalidIdentityToken: OpenIDConnect provider's HTTPS certificate doesn't match configured thumbprint\n\tstatus code: 400`

All praise due to this magician: https://marcincuber.medium.com/amazon-eks-with-oidc-provider-iam-roles-for-kubernetes-services-accounts-59015d15cb0c

> The reason it wasn’t working was the missing thumbprint for root CA. If you create OIDC provider for your EKS cluster using AWS Console, thumbprint automatically gets added. However, when you create it using terraform the list is empty.

Our EKS cluster was created via Terraform. Put another way, the OIDC provider I had associated with the cluster did not have the cluster CA thumbprint in its thumbprints_list.

Changing that in my OIDC provider's terraform configuration as per the above article solved this issue.
